docs: Use FLUO to update nodes (#1295)
kinvolk/lokomotive: docs/how-to-guides/auto-update-flatcar.md

---
title: Auto-updating Flatcar Container Linux
weight: 10
---

## Introduction

> Lokomotive supports Flatcar Container Linux only. Flatcar Linux Update Operator (FLUO) can keep the operating system updated every time there is a new Flatcar release. FLUO is a supported component of the Lokomotive's component stack. This document will show how to install it and disable its auto-update feature for specific nodes.

Reviewer: Maybe we can link to FLUO repository here?

Also, I'd add line breaks.

Reviewer: Also, "Lokomotive's component stack" sounds like a new term to me. Maybe we can just say "FLUO is one of the supported Lokomotive components"?

Reviewer: `This document` -> `This guide`.

## Prerequisites

> - A Lokomotive cluster accessible via `kubectl`.

Reviewer: It can actually be used on any K8s cluster running Flatcar.

## Steps

### Step 1: Disable auto-update on nodes

> If you want to update the Flatcar OS of particular nodes in a controlled fashion, there is a way to disable such updates. Disabling updates can come handy when the workloads run by these machines are storage related or ingress nodes, where you cannot afford the slightest loss of connection.

Reviewer: Ditto linebreaks.

Reviewer: >If you want to update the Flatcar OS of particular nodes in a controlled fashion, there is a way to disable such updates. 

It seems to me this sentence is incorrect. Shouldn't it be "If you **don't** want to update"?

We should also mention that disabling updates is not recommended.

Reviewer: Perhaps we could also document how to use hooks so nodes can be gracefully taken out from traffic etc.

Author: > If you want to update the Flatcar OS of particular nodes in a controlled fashion, there is a way to disable such updates.

This meant for manual updates. Why disabling updates is not recommended, until we provide a way to do it.

> Perhaps we could also document how to use hooks so nodes can be gracefully taken out from traffic etc.

Yes, but it is out of the scope of this PR. This PR concerns only with documenting the happy path updates. Once we have this foundation we can add more information on how to update it with nodes running 

- Networking apps (ingress, metallb)
- Storage apps (openebs, rook, self hosted storage)

Please annotate the nodes with the following annotation:

```bash
kubectl annotate node <node name> "flatcar-linux-update.v1.flatcar-linux.net/reboot-paused=true"
```

### Step 2: Configure FLUO

> Create a file named `fluo.lokocfg` with the following contents:

Reviewer: I'd make a file name an example, as it's not a hard requirement. Say: "Add FLUO component to your cluster configuration, for example by creating `fluo.lokocfg` file with the following content".

```tf
component "flatcar-linux-update-operator" {}
```

### Step 3: Install FLUO

Execute the following command to deploy the `flatcar-linux-update-operator` component:

```bash
lokoctl component apply flatcar-linux-update-operator
```

> Verify the pods in the `kube-system` namespace are in the `Running` state (this may take a few minutes):

Reviewer: In `reboot-coordinator` of course :)

```bash

> kubectl -n reboot-coordinator get pods

Reviewer 2: :smile:

```suggestion
kubectl -n kube-system get pods
```

---
title: Auto-updating Flatcar Container Linux
weight: 10
---

## Introduction

> Lokomotive supports Flatcar Container Linux only. [Flatcar Linux Update Operator
> (FLUO)](https://github.com/kinvolk/flatcar-linux-update-operator) can keep the operating system
> updated every time there is a new Flatcar release. FLUO is a supported Lokomotive component. This
> guide will show how to install it and disable its auto-update feature for specific nodes.

Reviewer: I'd add more context to introduction to make it more comprehensive. 
```suggestion
Right now Lokomotive only supports [Flatcar Container Linux](https://www.flatcar-linux.org/) as underlying operating system for nodes. While Flatcar can keep itself up to date, when running Kubernetes on top of it it is recommended to use [Flatcar Linux Update Operator
(FLUO)](https://github.com/kinvolk/flatcar-linux-update-operator) to perform updates to avoid rebooting too many cluster nodes at a time, which could cause the outage for services running on your cluster.

FLUO is a supported Lokomotive component.

This guide will show how to install it and disable its auto-update feature for specific nodes, which might be useful when you run services, which require special care before shutting down (e.g. storage clusters).
```

## Prerequisites

> - A Kubernetes cluster accessible via `kubectl`.

Reviewer: ```suggestion
- A Kubernetes cluster accessible via `kubectl` using Flatcar Container Linux for nodes.
```

## Steps

> ### Step 1: Disable auto-update on nodes

Reviewer: Hm, this is misleading. Looking at the header I thought we talk about disabling `locksmithd`. 

Perhaps we should add an optional step for it for non Lokomotive clusters?

Then, I'd change this title and perhaps mention FLUO here and maybe say something like "on sensitive nodes"? Because just saying `on nodes` seems to me like it referes to ALL nodes, which is incorrect.

Author: I don't think this document is a generic for anybody using FLUO anywhere. This is written with Lokomotive user's priority in mind. For a generic document the right place for it is in FLUO repo.

Reviewer: Well, this is for FLUO **component** IMO, which is part of Lokomotive. Also, Lokomotive itself disables `locksmithd`, but components can be used on all kinds of clusters.

Author: Yep I agree that FLUO can be used on all kinds of clusters. But we can't guarantee how a certain Lokomotive component will behave on a third-party cluster because there are assumptions made about the environment (in this case Lokomotive cluster). 

For example, we cannot guarantee component security on a cluster that does not enable PSP. Similarly we cannot guarantee network security to the machines because the third party cluster might not have something similar to calico globalnetworkpolicies the way we ship them.

Given these assumptions we can omit unnecessary details a Lokomotive user has to know. In this case they care about updating the OS that powers Lokomotive and they want a way to do it. They don't wanna know how what `locksmithd` is since we take care of it for them.

Reviewer: Well, I think treating components as cluster-agnostic, even when the compatibility is not guaranteed is something valuable. What do you think @iaguis?

If you want to update the Flatcar OS of particular nodes manually in a controlled fashion, there is
a way to disable automatic updates. Disabling updates can come handy when the workloads run by these
machines are storage related or ingress nodes, where you cannot afford the slightest loss of

> connection.

Reviewer: "slightest loss of connection." sounds odd to me here. Maybe we can use some simpler expression here? But I cannot come up with any suggestion.

Reviewer: Looks good, thanks!

Please annotate the nodes with the following annotation:

```bash
kubectl annotate node <node name> "flatcar-linux-update.v1.flatcar-linux.net/reboot-paused=true"
```

### Step 2: Configure FLUO

> Add the following content to your cluster configuration (`*.lokocfg` file):

Reviewer: ```suggestion
Add the following content to your cluster configuration (e.g. in `fluo.lokocfg` file):
```

```tf
component "flatcar-linux-update-operator" {}
```

### Step 3: Install FLUO

Execute the following command to deploy the `flatcar-linux-update-operator` component:

```bash
lokoctl component apply flatcar-linux-update-operator
```

Verify the pods in the `reboot-coordinator` namespace are in the `Running` state (this may take a few
minutes):

```bash
kubectl -n reboot-coordinator get pods

> ```

Reviewer: At the end, I'd add some testing commands like mentioned in https://github.com/kinvolk/flatcar-linux-update-operator#test so one can manually verify that the operator is functional.

## Prerequisites

- A Kubernetes cluster accessible via `kubectl` using Flatcar Container Linux for nodes.

## Steps

### Step 1: Disable auto-update on sensitive nodes

If you want to update the Flatcar OS of particular nodes manually in a controlled fashion, there is
a way to disable automatic updates. Disabling updates can come handy when the workloads run by these
machines are storage or ingress network related, where the applications cannot tolerate abrupt
reboot of node.

Please annotate the nodes with the following annotation:

```bash
kubectl annotate node <node name> "flatcar-linux-update.v1.flatcar-linux.net/reboot-paused=true"
```

### Step 2: Configure FLUO

Add the following content to your cluster configuration (e.g. in `fluo.lokocfg` file):

```tf
component "flatcar-linux-update-operator" {}
```

### Step 3: Install FLUO

Execute the following command to deploy the `flatcar-linux-update-operator` component:

```bash
lokoctl component apply flatcar-linux-update-operator
```

Verify the pods in the `reboot-coordinator` namespace are in the `Running` state (this may take a few
minutes):

```bash
kubectl -n reboot-coordinator get pods
```

### Step 4: Test installation

> To test that it is working, you can SSH into a node and trigger an update check by running

Reviewer: There is also 3rd option with annotating node objects we could mention: https://github.com/kinvolk/flatcar-linux-update-operator/issues/57

Author: Q: Why does it need both label and annotation? Just annotation or label won't do? I want to keep using only one way of doing stuff either label or annotation. Earlier in this doc (the one added by doc) we suggest adding annotation now we suggest adding both label and annotation. I wonder if both are needed. We use annotation in CI code as well.

Reviewer: IIRC one is produced by the agent and another one is produced by operator. Perhaps something to clarify in FLUO docs ;)

Author: So which one is the right to use?

Reviewer: Only the annotation I think. Label is only used for status.

---
title: Auto-updating Flatcar Container Linux
weight: 10
---

## Introduction

At the moment, Lokomotive supports only [Flatcar Container Linux](https://www.flatcar-linux.org/) as
an underlying operating system for nodes. While Flatcar can keep itself up to date, when running
Kubernetes on top of it, we recommend using [Flatcar Linux Update Operator
(FLUO)](https://github.com/kinvolk/flatcar-linux-update-operator) to perform updates to avoid
rebooting too many nodes at a time, which could cause the outage for services.

FLUO is a supported Lokomotive component.

This guide will show how to install it and disable its
auto-update feature for specific nodes, which might be useful when you run services, which require
special care before shutting down (e.g. storage clusters).

## Prerequisites

> - A Kubernetes cluster accessible via `kubectl` using Flatcar Container Linux for nodes.

Reviewer 3: If we assume Lokomotive, let's make it explicit :)
```suggestion
- A Lokomotive cluster accessible via `kubectl`.
```

## Steps

### Step 1: Disable auto-update of sensitive

If you want to update the Flatcar OS of particular nodes manually in a controlled fashion, there is
a way to disable automatic updates. Disabling updates can come handy when the workloads run by these
machines are storage or ingress network related, where the applications cannot tolerate abrupt
reboot of node.

Please annotate the nodes with the following annotation:

```bash
kubectl annotate node <node name> "flatcar-linux-update.v1.flatcar-linux.net/reboot-paused=true"
```

### Step 2: Configure FLUO

Add the following content to your cluster configuration (e.g. in `fluo.lokocfg` file):

```tf
component "flatcar-linux-update-operator" {}
```

### Step 3: Install FLUO

Execute the following command to deploy the `flatcar-linux-update-operator` component:

```bash
lokoctl component apply flatcar-linux-update-operator
```

> Verify the pods in the `reboot-coordinator` namespace are in the `Running` state (this may take a few

Reviewer 3: ```suggestion
Verify that pods in the `reboot-coordinator` namespace are in the `Running` state (this may take a few
```

minutes):

```bash
kubectl -n reboot-coordinator get pods
```

> 

Reviewer 3: I'm missing some explanation on how the cluster behaves now that FLUO is installed, describing that nodes will be updated automatically when an update is available and this will happen 1 by 1 with nodes that reboot for update being drained before that, and so on.

Author: Added a line here.

> ### Step 4: Test installation

Reviewer 3: I'd make this optional since it's potentially disrupting.

```suggestion
### Step 4: Test installation (optional)
```

---
title: Auto-updating Flatcar Container Linux
weight: 10
---

## Introduction

At the moment, Lokomotive supports only [Flatcar Container Linux](https://www.flatcar-linux.org/) as
the underlying operating system for nodes. While Flatcar can keep itself up to date, when running
Kubernetes on top of it, we recommend using the [Flatcar Linux Update Operator
(FLUO)](https://github.com/kinvolk/flatcar-linux-update-operator) to perform updates to avoid
rebooting too many nodes at a time which could cause a service outage.

FLUO is a supported Lokomotive component.

This guide will show how to install it and disable its
auto-update feature for specific nodes, which might be useful when you run services, which require
special care before shutting down (e.g. storage clusters).

> 

Reviewer: Suggestion: I still think this should be mentioned for completeness.
```suggestion

NOTE: If you want to use FLUO component on non-Lokomotive cluster, make sure your Flatcar nodes has `locksmithd.service` systemd unit disabled to avoid nodes rebooting on their own. On Lokomotive this is done out of the box.
```

## Prerequisites

- A Lokomotive cluster accessible via `kubectl`.

## Steps

### Step 1: Disable auto-update for sensitive nodes

If you want to update the particular nodes manually in a controlled fashion, there is a way to
disable automatic updates. Disabling updates can come in handy when the workloads run by these
machines are storage or ingress network related, where applications cannot tolerate the abrupt
reboot of node.

Please annotate the nodes as follows:

```bash
kubectl annotate node <node name> "flatcar-linux-update.v1.flatcar-linux.net/reboot-paused=true"
```

### Step 2: Configure FLUO

Add the following content to your cluster configuration (e.g. in `fluo.lokocfg` file):

```tf
component "flatcar-linux-update-operator" {}
```

### Step 3: Install FLUO

Execute the following command to deploy the `flatcar-linux-update-operator` component:

```bash
lokoctl component apply flatcar-linux-update-operator
```

Verify that pods in the `reboot-coordinator` namespace are running (this may take a
few minutes):

```bash
kubectl -n reboot-coordinator get pods
```

Now that you have installed FLUO, nodes without annotation

> `flatcar-linux-update.v1.flatcar-linux.net/reboot-paused=true` will be updated automatically when an
> update is available. One-by-one, the selected node is first drained before the reboot for an update.

Reviewer: Maybe re-phrase a bit to avoid using `update` word so many times.
```suggestion
`flatcar-linux-update.v1.flatcar-linux.net/reboot-paused=true` will be updated automatically when new version of Flatcar is available. One-by-one, the selected node is first drained before the reboot for an update.
```
